src/common/eeprom.c

```diff
@@ -2,228 +2,303 @@
+
+    version = eeprom_get_var(EEVAR_VERSION).ui16;
+    features = (version >= 4) ? eeprom_get_var(EEVAR_FEATURES).ui16 : 0;
+    if ((version >= EEPROM_FIRST_VERSION_CRC) && !eeprom_check_crc32())
+        defaults = 1;
+    else if ((version != EEPROM_VERSION) || (features != EEPROM_FEATURES)) {
+        if (eeprom_convert_from(version, features) == 0)
+            defaults = 1;
     }
-    //eeprom_print_vars();
+    if (defaults)
+        eeprom_defaults();
+    eeprom_print_vars();
     //eeprom_dump();
-    return ret;
+    return defaults;
 }
 
-// write default values to all variables
 void eeprom_defaults(void) {
-    uint8_t id;
-    for (id = 0; id < EE_VAR_CNT; id++) {
-        if (strcmp(eeprom_var_name[id], "EEVAR_LAN_HOSTNAME_START") == 0) {
-            eeprom_set_var(id, variant8_ui8('M'));
-            eeprom_set_var(id + 1, variant8_ui8('I'));
-            eeprom_set_var(id + 2, variant8_ui8('N'));
-            eeprom_set_var(id + 3, variant8_ui8('I'));
-            for (int id2 = id + 4; id2 < id + LAN_HOSTNAME_MAX_LEN; id2++) {
-                eeprom_set_var(id2, variant8_ui8(0));
-            }
-        } else {
-            eeprom_set_var(id, eeprom_var_default(id));
-        }
-    }
+    eeprom_vars_t vars = eeprom_var_defaults;
+    eeprom_lock();
+    // calculate crc32
+    vars.CRC32 = crc32_calc((uint32_t *)(&vars), (EEPROM_DATASIZE - 4) / 4);
+    // write data to eeprom
+    st25dv64k_user_write_bytes(EEPROM_ADDRESS, (void *)&vars, EEPROM_DATASIZE);
+    eeprom_unlock();
 }
 
 variant8_t eeprom_get_var(uint8_t id) {
     uint16_t addr;
     uint16_t size;
+    uint16_t data_size;
+    void *data_ptr;
     variant8_t var = variant8_empty();
-    if (id < EE_VAR_CNT) {
-        var.type = eeprom_map_v1[id];
-        addr = eeprom_var_addr(id);
+    if (id < EEPROM_VARCOUNT) {
+        eeprom_lock();
+        var = variant8_init(eeprom_map[id].type, eeprom_map[id].count, 0);
         size = eeprom_var_size(id);
-        st25dv64k_user_read_bytes(addr, &(var.ui32), size);
+        data_size = variant8_data_size(&var);
+        if (size == data_size) {
+            addr = eeprom_var_addr(id);
+            data_ptr = variant8_data_ptr(&var);
+            st25dv64k_user_read_bytes(addr, data_ptr, size);
+        } else {
+            _dbg("error");
+            //TODO:error
+        }
+        eeprom_unlock();
     }
     return var;
 }
 
 void eeprom_set_var(uint8_t id, variant8_t var) {
     uint16_t addr;
     uint16_t size;
-    if (id < EE_VAR_CNT) {
-        if (var.type == eeprom_map_v1[id]) {
-            addr = eeprom_var_addr(id);
+    uint16_t data_size;
+    void *data_ptr;
+    if (id < EEPROM_VARCOUNT) {
+        eeprom_lock();
+        if (var.type == eeprom_map[id].type) {
             size = eeprom_var_size(id);
-            st25dv64k_user_write_bytes(addr, &(var.ui32), size);
-            //			osDelay(5);
+            data_size = variant8_data_size(&var);
+            if (size == data_size) {
```

Comment:
In case of setting/getting a string: variant8_data_size is returning actual size of string (without '\0') and eeprom_var_size() is returning size of a string with a MAX length (shorter string is permitted). so this condition will never pass with strings.

Should be if(size >= data_size)

Reply:
This is right. You can set string variable using the variant_init with exact size, but this is an uncomfortable way and cannot be used with c++ interface only. I will change the condition to accept shorter data (this will be an exception only for strings - VARIANT8_PCHAR type).

Reply:
fixed in commit 9d92782